### ISSUE TITLE
Donate Ballista to Apache Arrow project

### ISSUE THREAD

**Issue by andygrove** [ballista-compute/ballista#646]
The Ballista project has recently reached a point where I believe that the basic architecture has been proven to work. The project has also suddenly become very popular and generated a lot of interest (more than 2k stars on GitHub). 

For these reasons, I think that the project has grown too large for me to continue maintaining as a personal project and I think it is now time to move the code to a foundation to ensure its continued success.

Given the deep dependencies on Apache Arrow (the core Arrow, DataFusion, and Parquet crates) and the fact that there is already some overlap between Arrow and Ballista committers, I believe that the obvious choice would be to donate the project to Apache Arrow.

Some of the benefits of donating the project to Arrow are:

- Ballista unit tests will be part of Arrow CI which means that any changes to Arrow or DataFusion APIs that Ballista depends on will also require that the corresponding Ballista code is updated as part of the same PR.
- DataFusion also needs a scheduler and it would probably make sense to push some parts of the Ballista scheduler down a level in the stack so that the same approach is used to scale across cores in DataFusion and to scale across nodes in Ballista.
- There is a team of committers that understand Arrow and DataFusion that can help with PR reviews so I will no longer be a bottleneck.
- Companies are more likely to commit resources to contributing to an Apache project compared to a personal project.

I am going to start a discussion on the Arrow mailing list to propose this idea as well.

**Comment by houqp** [comment]
Yeah, I would recommend a different repo as well so we can have our own release schedule. Another thing worth giving a try is not requiring jira issues on PRs like what Airlfow does, which helps reduce frictions for upstreaming contributions.

**Comment by Dandandan** [comment]
The first point, benefitting from easier integration with CI and applying changes in same PR as DataFusion / Arrow, we will not have when it moves to another repository.

I am 50/50 on the decision to put it in another repo as this first point would also bring a lot of benefits.

**Comment by edrevo** [comment]
Here's my wishlist, which comes from a perspective of someone with 0 knowledge of how Apache Arrow is organized internally as a community:

- If the plan is to push down the scheduler and the spill-to-disk functionality to DataFusion, I wonder at which point does is make sense to continue developing Ballista independently from DataFusion. Maybe merging both would be an option?
- I would also prefer an independent repo (ideally for Ballista + DataFusion). Maybe this repo could be moved to the apache org and renamed accordingly, and that way the GitHub stars would not be lost?
- I am (just slightly) worried that Jira is something that increases the entrance barrier for new contributors. I think GitHub issues + "good first issue" has worked rather well to attract new contributors lately. Not sure if moving to Jira is a must once you enter the ASF, but it would be great to keep having a very low-effort path for new contributors.

**Comment by houqp** [comment]
> The first point, benefitting from easier integration with CI and applying changes in same PR as DataFusion / Arrow

Good points. I agree with this, which is the biggest benefit for mono repo setup.

Perhaps another way to look at this is if Ballista ever becomes as successful as Spark, would it still make sense to have it hosted as a sub project within a message serialization format repo? Datafusion kind of makes sense to be part of arrow because it's a library. Ballista on the other hand is a full fledged application, which is soon going to even include a decent chunk of JS, CSS and HTML code :) It just seems like a bit too heavy to be included into arrow IMHO.

> Not sure if moving to Jira is a must once you enter the ASF, but it would be great to keep having a very low-effort path for new contributors.

It is not a must have because Airflow has made that off jira transition 2 years ago :)

**Comment by jorgecarleitao** [comment]
> Here's my wishlist, which comes from a perspective of someone with 0 knowledge of how Apache Arrow is organized internally as a community

@edrevo , to give some context; the arrow project is essentially a single git repository with:

* a format specification and its documentation
* 8+ different implementations of the specification at different levels of maturity (including C++, Java, Julia, Rust, JS), including documentation, tests, benchmarks, etc.
* 2+ language bindings (at least C++ to Python and R)
* 2 query engines (Grandiva and DataFusion)
* a data store (Plasma, deprecated but still there)
* a CI to test everything independently
* a CI that performs integration tests between all implementations.

Every PR requires an issue on the JIRA issue tracker. Creating a JIRA issue requires a user account and an admin to assign the user to a role.

Every implementation, query engine, binding, etc, is released under a single _major_ version every 3 months, whose latest is currently 3.0.0. This requires a formal vote by the Project management committee of 30 people (which @andygrove and I are part of). Implementations may release patch (not minor) without a vote, but they will still need to be release a new major every 3 months.

**Comment by jorgecarleitao** [comment]
I have been following this work, and it is truly amazing.

So, first of all, I want to thank all of you very much for making this happen, for keeping it open source, and for considering donating it to the Apache foundation and apache project in particular.

I am in favor of this for the same reasons that @andygrove presented. We do have a strong community on the Rust implementation that implemented much of the DataFusion. I would recommend the following:

1. DataFusion and Ballista reside on their own git repository (e.g. `arrow/ballista`)
2. Their release cycle and versioning are independent of the main arrow repo release cycle
3. DataFusion and Ballista's community decides whether they want to use JIRA

DataFusion would rely on a well defined Arrow and parquet crates' API. Ballista would host DataFusion as its query engine / library, which would allow:

* the Arrow repo to remain focused on maintaining the spec and aligned with the rest of the implementations.
* DataFusion features to be aligned with Ballista features (I understand that one is the multi-machine version of the other).

The rationale for the above is that whenever something is released on DataFusion (which hosts most of the physical ops), people will also want it quickly available on Ballista. Thus, having the two release cycles more closely related and independent of the arrow implementation's cycle is aligned. DataFusion does not have integration tests against other arrow implementations, and thus being the integration tests are not relevant here.

The risk here is that because arrow's release cycle is slow and on major only, DataFusion risks missing arrow features for some time. These deviations would require pointing to the repo's git and hash.

I would be happy to work with you in creating a comfortable repo for Ballista and DataFusion community to strive on.

The reason I want to bring the "how" already is that moving the Ballista code to the same repo would IMO be detrimental for its overall success. In particular, even I would be more willing to work on it than I currently am in the Arrow repo (for the same reasons that @houqp pointed out).

**Comment by alamb** [comment]
Haha I just posted the following to the arrow dev mailing list before reading this one. @jorgecarleitao did a better job than I explaining why pulling DataFusion out of the main arrow repo would be a good idea. I think we are in full agreement here. 


1. The scheduler and spill-to-disk/out of core operations sound very good to bring into DataFusion and many people would benefit
2. I think the arrow github project and the unified workflow process in particular is reaching its limits. Adding another cool, but non trivial project like Ballista will likely exacerbate the challenges even more. 
3. My sense is that the Rust arrow implementation is nearing feature completion (though we may still have one last big revamp, depending on @jorgecarleitao  plans) and so I expect breaking API changes there to slow down, lessening the value of keeping everything in the same rep. 
4.  What would you think about pulling DataFusion out of the arrow crate in the medium term (2-3 releases from now) and putting it into a new place (alongside Ballista)?

**Comment by andygrove** [comment]
Thanks for the great feedback @jorgecarleitao and @alamb.

@jorgecarleitao Would you mind posting a version of your feedback on the thread on the Arrow dev mailing list as well so that it gets visibility there?

I am all for DataFusion + Ballista being in a separate repo if that is an option.

Another thing I considered was proposing that this becomes a new top-level Apache project but my concerns with that are:

1. I don't know what the process is and perhaps Ballista would need to be much more mature before this can be considered?
2. It would probably be a lot of work for me to take on in my spare time (but I suppose it wouldn't just be me)

**Comment by andygrove** [comment]
Some updates from the conversation over on the Arrow mailing list. There is no reason why we can't have a separate release process for DataFusion/Ballista. Somebody just needs to do the work to make this happen. There is no need to have the code in a separate repo to achieve this.

**Comment by boazberman** [comment]
Contributing to Ballista has been great so far, and I think that what made it so is the face that there is very little bureaucracy, and it is very clear where things reside. I think that merging this product into a subfolder in Apache Arrow's monorepo would be its deathbed. 

The most demoralizing part of contributing to Ballista is that way to much things are overlapping with DataFusion, causing confusion as to what should go where, and how I as a contributor do one without knowing the other. The coupling is so great, that I feel dumb only getting familiar with one code base and not the other. 

I strongly believe that for Ballista & DataFusion to benefit the most from contributors and the community, a move to merge the two to a monorepo, separate from apache arrow, would be best.

Such move will also make it easier to create a unified api for local & distributed compute, which is one of the biggest pain points of the existing (python) data ecosystem - pyspark & pandas had drifted so much apart that new projects spawned (Dask) to fix just that. This project is in a unique position to fix that as well. We should not blow it without putting in some senciere effort.

**Comment by andygrove** [comment]
> I think that merging this product into a subfolder in Apache Arrow's monorepo would be its deathbed.

As a counterargument to that, people said similar things about DataFusion when I donated that to Apache Arrow and the project appears to be thriving now and I believe that it is in far better shape than it would have been if I had continued working on it as a personal project. DataFusion continues to evolve whether I am involved or not.

> The most demoralizing part of contributing to Ballista is that way to much things are overlapping with DataFusion, causing confusion as to what should go where, and how I as a contributor do one without knowing the other. The coupling is so great, that I feel dumb only getting familiar with one code base and not the other.

This sounds like a good argument for having DataFusion and Ballista in the same repository, which we would achieve by donating this project to Apache Arrow.

> I strongly believe that for Ballista & DataFusion to benefit the most from contributors and the community, a move to merge the two to a monorepo, separate from apache arrow, would be best.

I do see the appeal of having DataFusion and Ballista in the same repo, separate from the core Arrow project. I do see that the scope is larger. Donating Ballista to Arrow now does not mean that we can never move DataFusion and Ballista out into a separate repo later on (either under Arrow governance or as a new top-level project, depending on how things evolve).

> Such move will also make it easier to create a unified api for local & distributed compute, which is one of the biggest pain points of the existing (python) data ecosystem - pyspark & pandas had drifted so much apart that new projects spawned (Dask) to fix just that. This project is in a unique position to fix that as well. We should not blow it without putting in some senciere effort.

Again, this sounds like a good argument for having DataFusion and Ballista in the same repo, which we would achieve by donating the project to Apache Arrow. I agree that we should have tight coupling between DataFusion and Ballista so that we scale effortlessly both across cores on a laptop and across nodes in a cluster, using the same scheduler.

**Comment by andygrove** [comment]
> Every implementation, query engine, binding, etc, is released under a single _major_ version every 3 months, whose latest is currently 3.0.0. This requires a formal vote by the Project management committee of 30(?) people (which @andygrove and I are part of). Implementations may release patch (not minor) without a vote, but they still need to be released as a new major every 3 months.

I wanted to address this point since it is not completely accurate. There is a vote on major releases but it doesn't require everyone to vote. There is a minimum number of votes needed for the vote to pass (I forget the actual number) but as long as there are no -1 votes then it passes.

The reason that the release is only every 3 months currently is that we vote on the binary release, so that requires each team to create their binary artifacts and test them.

We recently voted to change this process for patch releases so that we vote on the source instead. This means the vote takes little effort and then each team is responsible for creating the binary artifacts (in their own time).

There have been discussions about moving to the same process for major releases and this would allow us to release more frequently. 

So, in summary, yes the release process isn't what we want today, but this is something that we are working on improving.

**Comment by houqp** [comment]
If we are to merge ballista and datafusion into the same arrow repo, will they be able to diverge and adopt their own versioning schedule? i.e. if there is a major version bump for ballista, let's say 5.x, do we need to bump the arrow crate to 5.x as well?

**Comment by alamb** [comment]
> If we are to merge ballista into the same arrow repo, will it be able to diverge and adopt its own versioning schedule? i.e. if there is a major version bump for ballista, let's say 5.x, do we need to bump the arrow crate to 5.x as well?

I assume the answer to this question is no. I envision datafusion / ballista being released under their own schedules and arrow remaining on its own schedule

ideally datafusion / ballista in my mind would begin to depend on the released version of Arrow rather than HEAD

**Comment by houqp** [comment]
That's great, that way we can achieve the goal of merging datafusion and ballista into the same code base while having a fully independent release schedule.

I agree that it would be ideal for datafusion and ballista to depend on released version of Arrow. Although if that's the plan, then it's best to donate it as a standalone Apache project, because IMO the biggest benefit of having them in the same repo is the following point initialy mentioned by Andy: `Ballista unit tests will be part of Arrow CI which means that any changes to Arrow or DataFusion APIs that Ballista depends on will also require that the corresponding Ballista code is updated as part of the same PR`.

Now my last suggestion would be for us to figure out a way to collaborate on the code base without requiring jira ticket on every PR :)

**Comment by alamb** [comment]
> Now my last suggestion would be for us to figure out a way to collaborate on the code base without requiring jira ticket on every PR :)

LOL -- everytime that topic is raised, it causes massive consternation. I think were we to put Ballista / Datafusion into their own repo we would have more flexibility about what tools were used 

> , because IMO the biggest benefit of having them in the same repo is the following point initialy mentioned by Andy: Ballista unit tests will be part of Arrow CI which means that any changes to Arrow or DataFusion APIs that Ballista depends on will also require that the corresponding Ballista code is updated as part of the same PR.

Yeah, on the other hand having it be more painful when the APIs change might help us reduce the breaking API changes and improve the experience for other Arrow users too

**Comment by boazberman** [comment]
As for the fear from breaking changes in arrow going unnoticed, it is fairly easy & free to test against master version of arrow daily.

As no one here will ever suggest to merge Ballista into the Rust repository to preemptively check for errors, I can't understand why doing this with Arrow is a good idea.

**Comment by andygrove** [comment]
I think the argument is more about having it in the same repo as DataFusion than having it in the same repo as Arrow (although Arrow is an overloaded term here).

There is (and should be) tight coupling between DataFusion and Ballista so it makes sense for them to be developed in the same repo and I think we are all in agreement on that point.

Where I think there is friction here is whether DataFusion and Ballista should be in their own repo separate from Arrow or not.

I think given the early nature of Ballista it would make sense to "incubate" it in the Arrow repo next to DataFusion and then revisit this conversation once Ballista is farther along.

**Comment by andygrove** [comment]
I just posted this message to the Arrow mailing list:

Thank you for all the responses so far. Based on this thread and the conversations happening in the Ballista project, I would say that the feedback is mostly positive and supportive of this donation, so I have started work on a PR [1] and will start a VOTE email thread once the PR is ready for review. Assuming that the vote passes, we will need to go through the IP clearance process.

To summarize the feedback so far (at least, my interpretation of it):

- There is consensus that it makes sense for DataFusion and Ballista to live in the same repository so that we can keep them tightly coupled and have queries scale across cores in DataFusion and across nodes in Ballista using the same scheduler.

- There is a desire to have a separate release process for DataFusion and Ballista that is not aligned with core Arrow releases, and it appears that there is no reason we cannot achieve this with Ballista living in the Arrow repo.

- There is a desire to have DataFusion and Ballista eventually live in a new repo, separate from core Arrow. There is no objection to still being under Arrow governance. If this donation goes through, I expect we will be discussing this point again at some point in the future, when Ballista is more mature.

- There is some concern that this will add an additional burden to current Arrow maintainers. I think we can reduce this burden by not requiring that Ballista depend on the HEAD version of DataFusion/Arrow i.e. we start doing some dependency management within the Rust project.

- There is some concern that being part of Arrow will make Ballista less attractive to contribute to due to the perceived "bureaucracy" of the ASF process, such as requiring JIRA tickets to be filed, and the (current) infrequent release cycles. I think this concern can be reduced over time. There were similar concerns when DataFusion was donated and that project seems to be thriving.

Thanks,

Andy.

[1] https://github.com/apache/arrow/pull/9723

**Comment by andygrove** [comment]
Hey everyone! The Apache Arrow PMC has passed a vote to accept the donation of Ballista.

I sent the following email out to those of you that I had email addresses for and who have code in the donation. I didn't have email address for @boaz-codota @j4ckcyw @pjmore @fitzoh @laplab so I am pinging you here.

The quickest way to get the PR merged is for everyone to sign the ICLA.

```
I'm excited to announce that the Apache Arrow PMC has passed a vote to accept the donation of the Ballista project!

You are receiving this email because you have contributed to Ballista in the past and your code may form part of the donation [1].

In order to make the IP clearance process go smoothly, I would greatly appreciate it if you could fill out an ICLA (https://www.apache.org/licenses/icla.pdf) and submit it to secretary@apache.org and cc me please (andygrove73@gmail.com)

The sooner we can merge the PR, the sooner development work can resume I'm excited for the future of Ballista and look forward to continuing to work with you.

Thanks,

Andy.

[1] https://github.com/apache/arrow/pull/9723
```

**Comment by andygrove** [comment]
@boaz-codota @j4ckcyw @hmakait @pjmore @zynaxsoft :point_up: I would appreciate it if you could submit an ICLA so that we can proceed with the donation. Thank you!

**Comment by andygrove** [comment]
also @hendrikmakait

**Comment by andygrove** [comment]
Thanks for everyone's patience with this. I think we have ICLAs now from just over half the contributors (thank you!). I am going to give this another day or so and will then proceed with a vote for the IP clearance process with the ASF. The project has been Apache-licensed since its inception so there should not be an issue with getting this passed.

**Comment by j4ckcyw** [comment]
just submitted today :)

**Comment by andygrove** [comment]
Last call for ICLAs! 

In order to make the IP clearance process go smoothly, I would greatly appreciate it if you could fill out an ICLA (https://www.apache.org/licenses/icla.pdf) and submit it to secretary@apache.org and cc me please (andygrove73@gmail.com)

@boaz-codota
@zynaxsoft
@dyqer
@meijies
@djKooks
@kyprifog
@laplab
@sieut
@phungleson
@vigsterkr
@rrichardson

Thank you in advance for your help with this.

**Comment by andygrove** [comment]
Here is the IP clearance submission on the ASF incubator web site:

https://incubator.apache.org/ip-clearance/arrow-rust-ballista.html

**Comment by andygrove** [comment]
To fulfill my obligations under the IP clearance process I would like to remind active committers that they are responsible for ensuring that a Corporate CLA is recorded if such is required to authorize their contributions under their individual CLA.
